python_modules/dagster/dagster/_core/execution/plan/execute_plan.py

```diff
@@ -35,106 +35,132 @@ def inner_plan_execution_iterator(
+        with ExitStack() as capture_stack:
+            # begin capturing logs for the whole process if this is a captured log manager
+            if isinstance(compute_log_manager, CapturedLogManager):
+                file_key = create_compute_log_file_key()
+                log_key = compute_log_manager.build_log_key_for_run(
+                    pipeline_context.run_id, file_key
+                )
+                try:
+                    log_context = capture_stack.enter_context(
+                        compute_log_manager.capture_logs(log_key)
+                    )
+                    yield DagsterEvent.capture_logs(
+                        pipeline_context, step_keys, log_key, log_context
+                    )
+                except Exception as e:
+                    yield DagsterEvent.engine_event(
+                        plan_context=pipeline_context,
+                        message="Exception while setting up compute log capture",
+                        event_specific_data=EngineEventData(
+                            error=serializable_error_info_from_exc_info(sys.exc_info())
+                        ),
+                    )
+                    log_capture_error = e
+
+            # It would be good to implement a reference tracking algorithm here to
+            # garbage collect results that are no longer needed by any steps
+            # https://github.com/dagster-io/dagster/issues/811
+            while not active_execution.is_complete:
+                step = active_execution.get_next_step()
+                step_context = cast(
+                    StepExecutionContext,
+                    pipeline_context.for_step(step, active_execution.get_known_state()),
+                )
+                step_event_list = []
+
+                missing_resources = [
+                    resource_key
+                    for resource_key in step_context.required_resource_keys
+                    if not hasattr(step_context.resources, resource_key)
+                ]
+                check.invariant(
+                    len(missing_resources) == 0,
+                    (
+                        "Expected step context for solid {solid_name} to have all required resources, but "
+                        "missing {missing_resources}."
+                    ).format(
+                        solid_name=step_context.solid.name, missing_resources=missing_resources
+                    ),
+                )
 
-            with ExitStack() as step_stack:
-                if not isinstance(compute_log_manager, CapturedLogManager):
-                    # capture all of the logs for individual steps
-                    log_capture_error = None
-                    try:
-                        step_stack.enter_context(
-                            pipeline_context.instance.compute_log_manager.watch(
-                                step_context.pipeline_run, step_context.step.key
+                with ExitStack() as step_stack:
```

Comment:
do we need another whole stack here versus just popping one off of a single exit stack?

Reply:
yeah, this is because we can't pop a single CM off the stack... it's either close or pop_all.